diff --git a/dist/index.cjs b/dist/index.cjs
--- a/dist/index.cjs
+++ b/dist/index.cjs
@@ -304 +304 @@
-			if (skipDuplicate && index && hasDeclaration(index, decl.prop, nextValue)) return;
+			if (!replace && skipDuplicate && index && hasDeclaration(index, decl.prop, nextValue)) return;
diff --git a/dist/index.mjs b/dist/index.mjs
--- a/dist/index.mjs
+++ b/dist/index.mjs
@@ -303 +303 @@
-			if (skipDuplicate && index && hasDeclaration(index, decl.prop, nextValue)) return;
+			if (!replace && skipDuplicate && index && hasDeclaration(index, decl.prop, nextValue)) return;
